### PR TITLE
POC: DataFrame.stack not including NA rows

### PR DESCRIPTION
**Pull request by rhshadrach**
Demonstration for #53515

- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by rhshadrach** [comment]
@jorisvandenbossche - there are some cases where DataFrame.stack sorts the result; I believe these are bugs. Other than this, the only tricky situation I've encountered is handling duplicates in the column that is being stacked. In this POC, I raise a ValueError instead. See #53761

File: pandas/core/frame.py [hunk 1]
```diff
@@ -9158,15 +9158,22 @@ def stack(self, level: IndexLabel = -1, dropna: bool = True, sort: bool = True):
...
-        else:
-            result = stack(self, level, dropna=dropna, sort=sort)
+        if (
+            isinstance(level, (tuple, list))
+            and not all(lev in self.columns.names for lev in level)
+            and not all(isinstance(lev, int) for lev in level)
+        ):
+            raise ValueError(
+                "level should contain all level names or all level "
+                "numbers, not a mixture of the two."
```
Comment (jorisvandenbossche) [inline pandas/core/frame.py:9170]: Is this a new restriction, or just moving up from the previous `stack_multiple` implementation?

(it certainly sounds as a good restriction, though)

File: pandas/core/reshape/reshape.py [hunk 2]
```diff
@@ -876,3 +881,110 @@ def _reorder_for_extension_array_stack(
...
+        ratio = 0
+
+    if len(level) < frame.columns.nlevels:
+        # concat column order may be different from dropping the levels
+        desired_columns = frame.columns._drop_level_numbers(drop_levnums).unique()
+        if not result.columns.equals(desired_columns):
+            result = result[desired_columns]
+
+    # Construct the correct MultiIndex by combining the frame's index and
+    # stacked columns.
```
Comment (jorisvandenbossche) [inline pandas/core/reshape/reshape.py:949]: You might be able to go back to your original simpler implementation, now that MultiIndex.append / concat performance has been improved (https://github.com/pandas-dev/pandas/pull/53697). 
Didn't check if your custom version here is still faster though, but that PR implemented my suggested improvement from profiling this stack use case.

File: pandas/core/reshape/reshape.py [hunk 3]
```diff
@@ -876,3 +881,110 @@ def _reorder_for_extension_array_stack(
     #  c0r1, c1r1, c2r1, ...]
     idx = np.arange(n_rows * n_columns).reshape(n_columns, n_rows).T.ravel()
     return arr.take(idx)
+
+
+def stack_v2(frame, level: list[int], dropna: bool = True, sort: bool = True):
```
Comment (jorisvandenbossche) [inline pandas/core/reshape/reshape.py:886]: The `sort` keyword is currently ignored in this implementation?

File: pandas/core/frame.py (hunk 1 above)
Comment (rhshadrach) [reply pandas/core/frame.py:9170]: No - this already exists in the current implementation.

File: pandas/core/reshape/reshape.py (hunk 2 above)
Comment (rhshadrach) [reply pandas/core/reshape/reshape.py:949]: I would have to guess that constructing a MultiIndex once is faster than constructing a MultiIndex many times and concat'ing them together. In any case, the complexity really arose from getting this to pass our suite of tests. The tests we have for stack/unstack are quite thorough.

**Comment by rhshadrach** [comment]
Closing in favor of #53921. @jorisvandenbossche - happy to continue any discussion here, or move over to #53921.

File: pandas/core/reshape/reshape.py (hunk 2 above)
Comment (jorisvandenbossche) [reply pandas/core/reshape/reshape.py:949]: Yes, indeed, just reusing the levels and repeating or tiling the existing codes should always be (at least a bit) faster than concatting

**Pull request closed without merging**